flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonScan.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes
+
+import org.apache.flink.api.common.functions.MapFunction
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.table.api.TableConfig
+import org.apache.flink.table.codegen.CodeGenerator
+import org.apache.flink.table.runtime.MapRunner
+import org.apache.flink.types.Row
+
+/**
+  * Common class for batch and stream scans.
+  */
+trait CommonScan {
+
+  /**
+    * We check if the input type is exactly the same as the internal row type.
+    * A conversion is necessary if types differ or object have to be unboxed
+    * (i.e. Date, Time, Timestamp need to be converted into their primitive equivalents).
+    */
+  private[flink] def needsConversion(
+      externalTypeInfo: TypeInformation[Any],
+      internalTypeInfo: TypeInformation[Row])
+    : Boolean = {
+
+    if (externalTypeInfo == internalTypeInfo) {
+      val rowTypeInfo = externalTypeInfo.asInstanceOf[RowTypeInfo]
+      var containsBoxedTypes = false
+      // TODO enable these lines for FLINK-5429
```

Comment:
Not sure if this would work as expected. Shouldn't we rather change `FlinkTypeFactory.toInternalRowType` to use `long`/`int` for date and time fields? If the types are changed, the internalTypeInfo (which is expected to be equal to the externalTypeInfo) would need to change too.